waspc/data/Generator/templates/react-app/src/actions/index.js

```diff
@@ -0,0 +1,87 @@
+import {
+  useMutation,
+  useQueryClient,
+} from 'react-query'
+
+export { configureQueryClient } from '../queryClient'
+
+export function useAction(actionFn, actionOptions) {
+  const queryClient = useQueryClient();
+
+  let mutationFn = actionFn
+  // TODO(filip): Do we allow all react query options or should we consider a list of allowed options?
+  let options = {}
+
+  if (actionOptions?.optimisticUpdates) {
+    const optimisticUpdatesConfig = makeOptimisticUpdatesConfig(actionOptions.optimisticUpdates)
+    mutationFn = makeOptimisticUpdateMutationFn(actionFn, optimisticUpdatesConfig)
+    options = makeOptimisticUpdateOptions(queryClient, optimisticUpdatesConfig)
+  }
+
+  return useMutation(mutationFn, options)
+}
+
+
+function makeOptimisticUpdatesConfig(optimisticUpdatesConfig) {
+  return optimisticUpdatesConfig.map(({ getQuery, ...rest }) => ({
+    getQuery: (item) => parseQueryKey(getQuery(item)),
+    ...rest,
+  }))
+}
+
+function makeOptimisticUpdateMutationFn(actionFn, optimisticUpdatesConfig) {
+  return function optimisticallyUpdateQueries(args) {
+    optimisticUpdatesConfig.forEach(({ getQuery }) => {
+      // how to make sure this is a query, a global query database?
+      const key = getQuery(args)
+      return actionFn.internal(args, [key])
+    })
+  }
+}
+
+function makeOptimisticUpdateOptions(queryClient, optimisticUpdatesConfig) {
```

Comment:
This entire function is an implementation of [react-query's recommended approach for optimistic updates](https://react-query.tanstack.com/guides/optimistic-updates).